src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -0,0 +1,772 @@
+        result.insert({transform_result, Chunk{}});
+
+    for (auto & [transform_result, cur_chunk] : result)
+    {
+        PaddedPODArray<UInt8> mask(chunk.getNumRows());
+        for (size_t i = 0; i < chunk.getNumRows(); ++i)
+        {
+            mask[i] = (transform_results[i] == transform_result);
+        }
+
+        for (size_t i = 0; i < chunk.getNumColumns(); ++i)
+        {
+            cur_chunk.addColumn(chunk.getColumns()[i]->filter(mask, 0));
+        }
+    }
+    return result;
+}
+
+IcebergStorageSink::IcebergStorageSink(
+    ObjectStoragePtr object_storage_,
+    ConfigurationPtr configuration_,
+    const std::optional<FormatSettings> & format_settings_,
+    const Block & sample_block_,
+    ContextPtr context_)
+    : SinkToStorage(sample_block_)
+    , sample_block(sample_block_)
+    , object_storage(object_storage_)
+    , context(context_)
+    , configuration(configuration_)
+    , format_settings(format_settings_)
+    , filename_generator(configuration_->getPath())
+{
+    configuration->update(object_storage, context, true, false);
+    auto log = getLogger("IcebergMetadata");
+    auto [last_version, metadata_path, compression_method]
+        = getLatestOrExplicitMetadataFileAndVersion(object_storage, configuration_, nullptr, context_, log.get());
+
+    filename_generator.setVersion(last_version + 1);
+    metadata = getMetadataJSONObject(metadata_path, object_storage, configuration, nullptr, context, log, compression_method);
+
+    partition_spec_id = metadata->getValue<Int64>("default-spec-id");
+    auto partitions_specs = metadata->getArray("partition-specs");
+
+    auto current_schema_id = metadata->getValue<Int64>(Iceberg::f_current_schema_id);
+    Poco::JSON::Object::Ptr current_schema;
+    auto schemas = metadata->getArray(Iceberg::f_schemas);
+    for (size_t i = 0; i < schemas->size(); ++i)
+    {
+        if (schemas->getObject(static_cast<UInt32>(i))->getValue<Int32>(Iceberg::f_schema_id) == current_schema_id)
+        {
+            current_schema = schemas->getObject(static_cast<UInt32>(i));
+        }
+    }
+    for (size_t i = 0; i < partitions_specs->size(); ++i)
+    {
+        auto current_partition_spec = partitions_specs->getObject(static_cast<UInt32>(i));
+        if (current_partition_spec->getValue<Int64>("spec-id") == partition_spec_id)
+        {
+            partititon_spec = current_partition_spec;
+            partitioner = ChunkPartitioner(current_partition_spec->getArray("fields"), current_schema, context_, sample_block_);
+            break;
+        }
+    }
+}
+
+void IcebergStorageSink::consume(Chunk & chunk)
+{
+    if (isCancelled())
+        return;
+    total_rows += chunk.getNumRows();
+    total_chunks_size += chunk.bytes();
+
+    std::unordered_map<ChunkPartitioner::PartitionKey, Chunk, ChunkPartitioner::PartitionKeyHasher> partition_result;
+    if (partitioner)
+    {
+        partition_result = partitioner->participateChunk(chunk);
```

Comment:
We have a code for MergeTree which solves similar problem https://github.com/scanhex12/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeDataWriter.cpp#L89-L155, Take look if not yet.